pkg/cloud/services/ec2/launchtemplate.go

```diff
@@ -689,14 +706,29 @@ func (s *Service) SDKToLaunchTemplate(d *ec2.LaunchTemplateVersion) (*expinfrav1
 	}
 
 	if v.UserData == nil {
-		return i, userdata.ComputeHash(nil), nil
+		return i, userdata.ComputeHash(nil), nil, nil
 	}
 	decodedUserData, err := base64.StdEncoding.DecodeString(*v.UserData)
 	if err != nil {
-		return nil, "", errors.Wrap(err, "unable to decode UserData")
+		return nil, "", nil, errors.Wrap(err, "unable to decode UserData")
+	}
+
+	var launchTemplateUserDataSecretKey *apimachinerytypes.NamespacedName
+	for _, tagSpecification := range v.TagSpecifications {
+		if tagSpecification.ResourceType != nil && *tagSpecification.ResourceType == ec2.ResourceTypeInstance {
+			for _, tag := range tagSpecification.Tags {
+				if tag.Key != nil && *tag.Key == infrav1.LaunchTemplateBootstrapDataSecret && tag.Value != nil && strings.Contains(*tag.Value, "/") {
+					parts := strings.SplitN(*tag.Value, "/", 2)
+					launchTemplateUserDataSecretKey = &apimachinerytypes.NamespacedName{
+						Namespace: parts[0],
+						Name:      parts[1],
+					}
+				}
+			}
+		}
 	}
 
-	return i, userdata.ComputeHash(decodedUserData), nil
+	return i, userdata.ComputeHash(decodedUserData), launchTemplateUserDataSecretKey, nil
```

Comment:
is it okay for `launchTemplateUserDataSecretKey` to be potentially nil here, if the above loop fails? Or should we return an error here?

Reply:
We must be backward-compatible and support cases where the tag is missing by mistake/bug/intervention. So `nil` is fine. I changed this line to explicit `nil` here, and added the `return` for the found case into the loop as you commented above.